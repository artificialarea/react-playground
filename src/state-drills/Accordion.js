import React from 'react';
import './Accordion.css';

// src: https://courses.thinkful.com/react-v1/checkpoint/10#assignment
// alt solution: https://github.com/tomatau/state-drills/tree/accordion/src/accordion


class Accordion extends React.Component {
  state = {
    currentLi: null,
    isExposed: false,
  }

  static defaultProps = {
    sections: [],
    isExposed: false,
  }

  handleButtonClick = (index) => {
    this.setState({
      currentLi: index,
      isExposed: !this.state.isExposed, // which toggles boolean
    })
  }

  renderLis() {
    return this.props.sections.map((section, index) => {
      return (
        <li key={index} >
          <button onClick={() => this.handleButtonClick(index)}>{section.title}</button>
          <p>{this.state.currentLi === index  && this.state.isExposed ? section.content : ""}</p>
        </li>
      )
    });
  }

  render() {
    return (
      <div>
        <h2>Accordion</h2>
        <ul>
          {/* {!!this.props.sections.length && this.renderLis()} */}
          {this.props.sections.length ? this.renderLis() : <li/>}
        </ul>
      </div>
    )
  }
}

export default Accordion;

/*
export default class App extends React.Component {
    constructor(props) {
        super(props);
        this.state = { isLiked: false, name: "Mickey Mouse" };
    }

    _toggleLikeState = () => {
        this.setState({
            isLiked: !this.state.isLiked
        });
    }

    render() {
        const { name, isLiked } = this.state;
        console.log(name);
        return (
            <div className="parent">
                <ChangeMeButton name={name} onToggleClick={() => this._toggleLikeState()} />
                <span>{ isLiked ? "yay!": "nay!" }</span>
            </div>
        );
    }
}
*/