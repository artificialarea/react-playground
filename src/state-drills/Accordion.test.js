import React from 'react';
import ReactDOM from 'react-dom';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import STORE from './accordion-store'

import Accordion from './Accordion';

describe('<Accordion />', () => {

  it('renders without errors', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Accordion />, div)
    ReactDOM.unmountComponentAtNode(div)
  });

  it('renders empty li when section prop not supplied', () => {
    const wrapper = shallow(<Accordion />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('component renders no sections as active by default', () => {
    const wrapper = shallow(<Accordion sections={STORE.sections} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('opens when a section is clicked', () => {
    const wrapper = shallow(<Accordion sections={STORE.sections} />);
    wrapper.find('button').at(0).simulate('click');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('component opens and closes sections with multiple clicks', () => {
    const wrapper = shallow(<Accordion sections={STORE.sections} />);
    
    wrapper.find('button').at(0).simulate('click');
    wrapper.find('button').at(2).simulate('click');

    console.log('>>> WRAPPER <<<');
    console.log(wrapper.debug());

    expect(toJson(wrapper)).toMatchSnapshot();
  });

});