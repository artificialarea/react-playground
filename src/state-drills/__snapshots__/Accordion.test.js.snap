// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> component opens and closes sections with multiple clicks 1`] = `
<div>
  <h2>
    Accordion
  </h2>
  <ul>
    <li
      key="0"
    >
      <button
        onClick={[Function]}
      >
        Section 1
      </button>
      <p />
    </li>
    <li
      key="1"
    >
      <button
        onClick={[Function]}
      >
        Section 2
      </button>
      <p />
    </li>
    <li
      key="2"
    >
      <button
        onClick={[Function]}
      >
        Section 3
      </button>
      <p />
    </li>
  </ul>
</div>
`;

exports[`<Accordion /> component renders no sections as active by default 1`] = `
<div>
  <h2>
    Accordion
  </h2>
  <ul>
    <li
      key="0"
    >
      <button
        onClick={[Function]}
      >
        Section 1
      </button>
      <p />
    </li>
    <li
      key="1"
    >
      <button
        onClick={[Function]}
      >
        Section 2
      </button>
      <p />
    </li>
    <li
      key="2"
    >
      <button
        onClick={[Function]}
      >
        Section 3
      </button>
      <p />
    </li>
  </ul>
</div>
`;

exports[`<Accordion /> opens when a section is clicked 1`] = `
<div>
  <h2>
    Accordion
  </h2>
  <ul>
    <li
      key="0"
    >
      <button
        onClick={[Function]}
      >
        Section 1
      </button>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit.
      </p>
    </li>
    <li
      key="1"
    >
      <button
        onClick={[Function]}
      >
        Section 2
      </button>
      <p />
    </li>
    <li
      key="2"
    >
      <button
        onClick={[Function]}
      >
        Section 3
      </button>
      <p />
    </li>
  </ul>
</div>
`;

exports[`<Accordion /> renders empty li when section prop not supplied 1`] = `
<div>
  <h2>
    Accordion
  </h2>
  <ul>
    <li />
  </ul>
</div>
`;
